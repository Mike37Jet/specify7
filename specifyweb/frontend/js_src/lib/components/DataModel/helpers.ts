import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import type {
  AnySchema,
  AnyTree,
  SerializedModel,
  SerializedResource,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import { parseResourceUrl, resourceToJson } from './resource';
import { strictGetModel } from './schema';
import type { Tables } from './types';
import { addMissingFields } from './addMissingFields';

/** Like resource.toJSON(), but keys are converted to camel case */
export const serializeResource = <SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA> | SpecifyResource<SCHEMA>
): SerializedResource<SCHEMA> =>
  serializeModel<SCHEMA>(
    typeof resource.toJSON === 'function'
      ? resourceToJson(resource as SpecifyResource<SCHEMA>)
      : (resource as SerializedModel<SCHEMA>),
    (resource as SpecifyResource<SCHEMA>)?.specifyModel?.name
  );

export const specialFields = new Set([
  'id',
  'resource_uri',
  'recordset_info',
  '_tableName',
]);

/** Recursive helper for serializeResource */
function serializeModel<SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA>,
  tableName?: keyof Tables
): SerializedResource<SCHEMA> {
  const model = strictGetModel(
    defined(
      (tableName as SCHEMA['tableName']) ??
        resource._tableName ??
        parseResourceUrl((resource.resource_uri as string) ?? '')?.[0],
      `Unable to serialize resource because table name is unknown.` +
        (process.env.NODE_ENV === 'test'
          ? `\nMake sure your test file calls requireContext();`
          : '')
    )
  );
  const fields = [...model.fields.map(({ name }) => name), model.idField.name];

  return addMissingFields(
    model.name,
    Object.fromEntries(
      Object.entries(resource).map(([lowercaseFieldName, value]) => {
        let camelFieldName = fields.find(
          (fieldName) => fieldName.toLowerCase() === lowercaseFieldName
        );
        if (camelFieldName === undefined) {
          camelFieldName = lowercaseFieldName;
          if (!specialFields.has(lowercaseFieldName))
            console.warn(
              `Trying to serialize unknown field ${lowercaseFieldName} for table ${model.name}`,
              resource
            );
        }
        if (
          typeof value === 'object' &&
          value !== null &&
          !specialFields.has(camelFieldName)
        ) {
          const field = model.getField(lowercaseFieldName);
          const tableName =
            field === undefined || !field.isRelationship
              ? undefined
              : field.relatedModel.name;
          return [
            camelFieldName,
            Array.isArray(value)
              ? value.map((value) =>
                  serializeModel(
                    value as unknown as SerializedModel<SCHEMA>,
                    tableName
                  )
                )
              : serializeModel(value as SerializedModel<AnySchema>, tableName),
          ];
        } else return [camelFieldName, value];
      })
    )
  ) as SerializedResource<SCHEMA>;
}

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyModel.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyModel.name === tableName ? resource : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyModel.name) ? resource : undefined;
