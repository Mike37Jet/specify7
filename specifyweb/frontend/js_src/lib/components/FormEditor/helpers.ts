import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AttachmentOwnerBaseBusRules',
  'AttachmentReferenceBaseBusRules',
  'AuthorBusRules',
  'BaseTreeBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventAttributeBusRules',
  'CollectingEventAuthorizationBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectionSetupBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeaccessionBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DisposalAgentBusRules',
  'DisposalBusRules',
  'DisposalPreparationBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeInBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeoCoordDetailBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'InstitutionBusRules',
  'JournalBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'LocalityDetailBusRules',
  'PaleoContextBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'ReferenceWorkBusRules',
  'RepositoryAgreementBusRules',
  'ShipmentBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TableSearchResults',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeableSearchQueryBuilder',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
