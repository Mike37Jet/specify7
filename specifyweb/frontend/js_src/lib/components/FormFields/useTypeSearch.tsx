import React from 'react';

import { useAsyncState } from '../../hooks/useAsyncState';
import { queryText } from '../../localization/query';
import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import { getParsedAttribute } from '../../utils/utils';
import { formatList } from '../Atoms/Internationalization';
import type { LiteralField, Relationship } from '../DataModel/specifyField';
import type { SpecifyModel } from '../DataModel/specifyModel';
import { load } from '../InitialContext';
import { formatUrl } from '../Router/queryString';
import { columnToFieldMapper } from './parseSelect';
import type { TypeSearch } from './queryComboBoxUtils';

const typeSearches = load<Element>(
  formatUrl('/context/app.resource', { name: 'TypeSearches' }),
  'text/xml'
);

export function useTypeSearch(
  initialTypeSearch: Element | string | undefined,
  field: LiteralField | Relationship | undefined,
  initialRelatedModel: SpecifyModel | undefined
): TypeSearch | false | undefined {
  const [typeSearch] = useAsyncState<TypeSearch | false>(
    React.useCallback(async () => {
      const relatedModel =
        initialRelatedModel ??
        (field?.isRelationship === true ? field.relatedModel : undefined);
      if (relatedModel === undefined) return false;

      const typeSearch =
        typeof initialTypeSearch === 'object'
          ? initialTypeSearch
          : (await typeSearches).querySelector(
              typeof initialTypeSearch === 'string'
                ? `[name="${initialTypeSearch}"]`
                : `[tableid="${relatedModel.tableId}"]`
            );
      if (typeSearch === undefined) return false;

      const rawSearchFieldsNames =
        typeSearch === null
          ? []
          : getParsedAttribute(typeSearch, 'searchField')
              ?.split(',')
              .map(f.trim)
              .map(
                typeof typeSearch?.textContent === 'string' &&
                  typeSearch.textContent.trim().length > 0
                  ? columnToFieldMapper(typeSearch.textContent)
                  : f.id
              ) ?? [];
      const searchFields = rawSearchFieldsNames.map((searchField) =>
        relatedModel.strictGetField(searchField)
      );

      const fieldTitles = searchFields.map((field) =>
        filterArray([
          field.model === relatedModel ? undefined : field.model.label,
          field.label,
        ]).join(' / ')
      );

      return {
        title: queryText('queryBoxDescription', formatList(fieldTitles)),
        searchFields,
        relatedModel,
        dataObjectFormatter:
          typeSearch?.getAttribute('dataObjFormatter') ?? undefined,
      };
    }, [initialTypeSearch, field, initialRelatedModel]),
    false
  );
  return typeSearch;
}
