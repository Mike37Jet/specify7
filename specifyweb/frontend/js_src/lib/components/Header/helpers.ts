export function formatDate(date: Date) {
  const year = date.getFullYear();
  const month = date.getMonth().toString();
  const day = date.getDate().toString();
  const hours = date.getHours().toString();
  const minutes = date.getMinutes().toString();
  const seconds = date.getSeconds().toString();

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}
