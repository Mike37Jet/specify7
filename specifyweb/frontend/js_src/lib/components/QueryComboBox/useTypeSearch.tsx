import React from 'react';

import { useAsyncState } from '../../hooks/useAsyncState';
import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { LiteralField, Relationship } from '../DataModel/specifyField';
import type { SpecifyModel } from '../DataModel/specifyModel';
import { load } from '../InitialContext';
import { formatUrl } from '../Router/queryString';
import { xmlToSpec } from '../Syncer/xmlUtils';
import type { TypeSearch } from './spec';
import { typeSearchesSpec } from './spec';

export const typeSearches = f
  .all({
    xml: load<Element>(
      formatUrl('/context/app.resource', { name: 'TypeSearches' }),
      'text/xml'
    ),
    schema: import('../DataModel/schema').then(
      async ({ fetchContext }) => fetchContext
    ),
  })
  .then(({ xml }) =>
    filterArray(xmlToSpec(xml, typeSearchesSpec()).typeSearches)
  );

export function useTypeSearch(
  initialTypeSearch: TypeSearch | string | undefined,
  field: LiteralField | Relationship,
  initialRelatedModel: SpecifyModel | undefined
): TypeSearch | false | undefined {
  const relatedModel =
    initialRelatedModel ??
    (field?.isRelationship ? field.relatedModel : undefined) ??
    (typeof initialTypeSearch === 'object'
      ? initialTypeSearch.table
      : undefined);
  const [typeSearch] = useAsyncState<TypeSearch | false>(
    React.useCallback(() => {
      if (typeof initialTypeSearch === 'object') return initialTypeSearch;
      else if (relatedModel === undefined) return false;
      return typeSearches.then(
        (typeSearches) =>
          typeSearches.find(({ name }) => name === initialTypeSearch) ??
          typeSearches.find(({ table }) => table === relatedModel)
      );
    }, [initialTypeSearch, relatedModel]),
    false
  );
  return typeSearch;
}
