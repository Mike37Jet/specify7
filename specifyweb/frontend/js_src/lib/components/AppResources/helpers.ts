import type { AppResourcesConformation } from './Aside';
import type { AppResources, AppResourcesTree } from './hooks';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { f } from '../../utils/functools';
import type { SpecifyModel } from '../DataModel/specifyModel';
import type { RA } from '../../utils/types';
import { SerializedResource } from '../DataModel/helperTypes';

export const getAppResource = (
  resources: AppResources,
  model: SpecifyModel<SpAppResource | SpViewSetObj>,
  resourceId: number | undefined
): SerializedResource<SpAppResource | SpViewSetObj> | false | undefined =>
  typeof resourceId === 'number'
    ? (
        (model.name === 'SpAppResource'
          ? resources.appResources
          : resources.viewSets) as RA<SerializedResource<SpAppResource>>
      ).find(({ id }) => id === resourceId) ?? false
    : undefined;

export type AppResourceMode = 'appResources' | 'viewSets';

export const getAppResourceMode = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): AppResourceMode =>
  resource._tableName === 'SpAppResource' ? 'appResources' : 'viewSets';

export const getAppResourceCount = (
  resourcesTree: AppResourcesTree[number]
): number =>
  resourcesTree.appResources.length +
  resourcesTree.viewSets.length +
  f.sum(resourcesTree.subCategories.map(getAppResourceCount));

export const buildAppResourceConformation = (
  resourcesTree: AppResourcesTree
): RA<AppResourcesConformation> =>
  resourcesTree
    .filter(
      ({ subCategories, appResources, viewSets }) =>
        subCategories.length > 0 ||
        appResources.length > 0 ||
        viewSets.length > 0
    )
    .map(({ key, subCategories }) => ({
      key,
      children: buildAppResourceConformation(subCategories),
    }));
