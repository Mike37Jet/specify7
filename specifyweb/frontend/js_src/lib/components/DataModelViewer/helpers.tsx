import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { schemaText } from '../../localization/schema';
import { syncFieldFormat } from '../../utils/fieldFormat';
import { f } from '../../utils/functools';
import { resolveParser } from '../../utils/parser/definitions';
import { ensure, IR, RA, RR } from '../../utils/types';
import { formatNumber } from '../Atoms/Internationalization';
import { getField } from '../DataModel/helpers';
import { schema } from '../DataModel/schema';
import { TableIcon } from '../Molecules/TableIcon';

export type Value =
  | number
  | string
  | readonly [number | string | undefined, JSX.Element]
  | undefined;
export type Row<SHAPE extends IR<Value>> = SHAPE;

const parser = f.store(() =>
  resolveParser(
    {},
    {
      type: 'java.lang.Boolean',
    }
  )
);

export const booleanFormatter = (value: boolean): string =>
  syncFieldFormat(undefined, parser(), value);

export const tableColumns = f.store(
  () =>
    ({
      name: getField(schema.models.SpLocaleContainer, 'name').label,
      label: schemaText.fieldLabel(),
      isSystem: getField(schema.models.SpLocaleContainer, 'isSystem').label,
      isHidden: getField(schema.models.SpLocaleContainer, 'isHidden').label,
      tableId: schemaText.tableId(),
      fieldCount: schemaText.fieldCount(),
      relationshipCount: schemaText.relationshipCount(),
    } as const)
);

export const getTables = () =>
  ensure<RA<Row<RR<keyof ReturnType<typeof tableColumns>, Value>>>>()(
    Object.values(schema.models).map(
      (model) =>
        ({
          name: [
            model.name.toLowerCase(),
            <>
              <TableIcon label={false} name={model.name} />
              {model.name}
            </>,
          ],
          label: model.label,
          isSystem: booleanFormatter(model.isSystem),
          isHidden: booleanFormatter(model.isHidden),
          tableId: [
            model.tableId,
            <span className="flex w-full justify-end tabular-nums">
              {model.tableId}
            </span>,
          ],
          fieldCount: [
            model.fields.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.fields.length)}
            </span>,
          ],
          relationshipCount: [
            model.relationships.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.relationships.length)}
            </span>,
          ],
        } as const)
    )
  );

export function DataModelRedirect(): null {
  const { tableName = '' } = useParams();
  const navigate = useNavigate();
  React.useEffect(() => {
    navigate(`/specify/data-model/#${tableName}`, { replace: true });
  }, []);
  return null;
}
