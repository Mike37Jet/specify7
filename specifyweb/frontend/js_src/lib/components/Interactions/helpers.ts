import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA } from '../../utils/types';

export type Preparations = RA<{
  readonly catalogNumber: string;
  readonly taxon: string;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
}>;

export type PreparationRow = readonly [
  string,
  string,
  number,
  string,
  number,
  string | null,
  string | null,
  string | null,
  string
];
export const getPrepsAvailableForLoanRs = async (recordSetId: number) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      headers: { Accept: 'application/json' },
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
