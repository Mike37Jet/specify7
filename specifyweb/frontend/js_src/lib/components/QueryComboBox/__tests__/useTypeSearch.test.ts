import { requireContext } from '../../../tests/helpers';
import { typeSearches } from '../useTypeSearch';

requireContext();

test('type searches are fetched and parsed correctly', async () =>
  expect(typeSearches).resolves.toMatchSnapshot());
