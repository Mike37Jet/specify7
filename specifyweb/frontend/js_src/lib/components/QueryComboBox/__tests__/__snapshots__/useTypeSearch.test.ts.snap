// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type searches are fetched and parsed correctly 1`] = `
[
  {
    "displayFields": [
      "accessionNumber",
    ],
    "format": "%s",
    "formatter": "",
    "name": "AccessionCO",
    "searchFields": [
      [
        "[literalField accessionNumber]",
      ],
    ],
    "table": "[table Accession]",
    "title": "Searched fields: Accession #",
  },
  {
    "displayFields": [
      "lastName",
      "firstName",
    ],
    "format": "%s, %s",
    "formatter": "Agent",
    "name": "Agent",
    "searchFields": [
      [
        "[literalField lastName]",
      ],
    ],
    "table": "[table Agent]",
    "title": "Searched fields: Last Name",
  },
  {
    "displayFields": [
      "lastName",
      "firstName",
    ],
    "format": "%s, %s",
    "formatter": "Agent",
    "name": "AgentSSPU",
    "searchFields": [
      [
        "[literalField lastName]",
      ],
    ],
    "table": "[table Agent]",
    "title": "Searched fields: Last Name",
  },
  {
    "displayFields": [
      "appraisalNumber",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Appraisal",
    "searchFields": [
      [
        "[literalField appraisalNumber]",
      ],
    ],
    "table": "[table Appraisal]",
    "title": "Searched fields: Appraisal Number",
  },
  {
    "displayFields": [
      "schemeName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "AutoNumberingScheme",
    "searchFields": [
      [
        "[literalField schemeName]",
      ],
    ],
    "table": "[table AutoNumberingScheme]",
    "title": "Searched fields: Schema Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "BioStrat",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table GeologicTimePeriod]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "schemeName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "CatAutoNumberingScheme",
    "searchFields": [
      [
        "[literalField schemeName]",
      ],
    ],
    "table": "[table AutoNumberingScheme]",
    "title": "Searched fields: Schema Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "ChronosStrat",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table GeologicTimePeriod]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "stationFieldNumber",
      "startDate",
      "loc.localityName",
      "geo.fullName",
    ],
    "format": "%s; %s; %s; %s",
    "formatter": "",
    "name": "CollectingEvent",
    "searchFields": [
      [
        "[literalField stationFieldNumber]",
      ],
    ],
    "table": "[table CollectingEvent]",
    "title": "Searched fields: Field Number",
  },
  {
    "displayFields": [
      "text1",
      "text2",
      "number1",
    ],
    "format": "%s:%s:%s",
    "formatter": "CollectingTrip",
    "name": "CollectingTrip",
    "searchFields": [
      [
        "[literalField collectingTripName]",
      ],
    ],
    "table": "[table CollectingTrip]",
    "title": "Searched fields: Trip Name",
  },
  {
    "displayFields": [
      "catalogNumber",
    ],
    "format": "",
    "formatter": "CollectionObject",
    "name": "CollectionObject",
    "searchFields": [
      [
        "[literalField catalogNumber]",
      ],
    ],
    "table": "[table CollectionObject]",
    "title": "Searched fields: Cat #",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Container",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table Container]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "formatter": "",
    "name": "ContainerOwner",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table Container]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "primerDesignator",
    ],
    "format": "%s",
    "formatter": "",
    "name": "DNAPrimer",
    "searchFields": [
      [
        "[literalField primerDesignator]",
      ],
    ],
    "table": "[table DNAPrimer]",
    "title": "Searched fields: Primer Designator",
  },
  {
    "displayFields": [
      "deaccessionNumber",
    ],
    "format": "%s",
    "formatter": "Deaccession",
    "name": "Deaccession",
    "searchFields": [
      [
        "[literalField deaccessionNumber]",
      ],
    ],
    "table": "[table Deaccession]",
    "title": "Searched fields: Deacc #",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Discipline",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table Discipline]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "disposalNumber",
    ],
    "format": "%s",
    "formatter": "Disposal",
    "name": "Disposal",
    "searchFields": [
      [
        "[literalField disposalNumber]",
      ],
    ],
    "table": "[table Disposal]",
    "title": "Searched fields: disposalNumber",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Division",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table Division]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "title",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Exsiccata",
    "searchFields": [
      [
        "[literalField title]",
      ],
    ],
    "table": "[table Exsiccata]",
    "title": "Searched fields: Title",
  },
  {
    "displayFields": [
      "fnbp.pageNumber",
      "fnb.name",
    ],
    "format": "%s - %s",
    "formatter": "FieldNotebookPage",
    "name": "FieldNotebookPage",
    "searchFields": [
      [
        "[literalField pageNumber]",
      ],
    ],
    "table": "[table FieldNotebookPage]",
    "title": "Searched fields: Page Number",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "Geography",
    "name": "Geography",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table Geography]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "GeologicTimePeriod",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table GeologicTimePeriod]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "journalName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Journal",
    "searchFields": [
      [
        "[literalField journalName]",
      ],
    ],
    "table": "[table Journal]",
    "title": "Searched fields: Journal Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "LithoStrat",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table LithoStrat]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "localityName",
      "geo.fullName",
    ],
    "format": "%s;  %s",
    "formatter": "Locality",
    "name": "Locality",
    "searchFields": [
      [
        "[literalField localityName]",
      ],
    ],
    "table": "[table Locality]",
    "title": "Searched fields: Locality Name",
  },
  {
    "displayFields": [
      "viewName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "MorphBankView",
    "searchFields": [
      [
        "[literalField viewName]",
      ],
    ],
    "table": "[table MorphBankView]",
    "title": "Searched fields: View Name",
  },
  {
    "displayFields": [
      "paleoContextName",
      "cs.fullName",
      "cse.fullName",
      "ls.fullName",
    ],
    "format": "%s, %s, %s, %s",
    "formatter": "PaleoContext",
    "name": "PaleoContext",
    "searchFields": [
      [
        "[literalField paleoContextName]",
      ],
      [
        "[relationship chronosStrat]",
        "[literalField fullName]",
      ],
      [
        "[relationship lithoStrat]",
        "[literalField fullName]",
      ],
    ],
    "table": "[table PaleoContext]",
    "title": "Searched fields: Context Name, Chronostratigraphy / Full Name, and Lithostratigraphy / Full Name",
  },
  {
    "displayFields": [
      "permitNumber",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Permit",
    "searchFields": [
      [
        "[literalField permitNumber]",
      ],
    ],
    "table": "[table Permit]",
    "title": "Searched fields: Permit Number",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "formatter": "",
    "name": "PrepType",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table PrepType]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "title",
    ],
    "format": "%s",
    "formatter": "",
    "name": "ReferenceWork",
    "searchFields": [
      [
        "[literalField title]",
      ],
    ],
    "table": "[table ReferenceWork]",
    "title": "Searched fields: Title",
  },
  {
    "displayFields": [
      "repositoryAgreementNumber",
    ],
    "format": "%s",
    "formatter": "",
    "name": "RepositoryAgreement",
    "searchFields": [
      [
        "[literalField repositoryAgreementNumber]",
      ],
    ],
    "table": "[table RepositoryAgreement]",
    "title": "Searched fields: RA Number",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "",
    "name": "Storage",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table Storage]",
    "title": "Searched fields: Name",
  },
  {
    "displayFields": [
      "fullName",
    ],
    "format": "%s",
    "formatter": "Taxon",
    "name": "Taxon",
    "searchFields": [
      [
        "[literalField fullName]",
      ],
    ],
    "table": "[table Taxon]",
    "title": "Searched fields: Full Name",
  },
  {
    "displayFields": [
      "name",
    ],
    "format": "%s",
    "formatter": "",
    "name": "TaxonTreeDefItem",
    "searchFields": [
      [
        "[literalField name]",
      ],
    ],
    "table": "[table TaxonTreeDefItem]",
    "title": "Searched fields: Name",
  },
]
`;
